spec/ssl-test_spec.rb

```diff
@@ -176,6 +183,41 @@
       expect(valid).to eq(true)
       expect(cert).to be_a OpenSSL::X509::Certificate
     end
+
+    context 'when specifying a proxy' do
+      let(:proxy_thread) do
+        Thread.new do
+          dev_null = WEBrick::Log::new("/dev/null", 7)
+          $proxy = WEBrick::HTTPProxyServer.new Port: 8080,  :Logger => dev_null, :AccessLog => []
+          $proxy.start
+        end
+      end
+
+      context 'when the proxy is active' do
+        it 'uses the provided http proxy' do
+          proxy_thread
+          sleep 0.1 # wait for the proxy to start!
+
+          allow($proxy).to receive(:do_GET).and_call_original
```

Comment:
Thanks for the assertion, quick note: I believe you can do it in a single line with:
```suggestion
          expect($proxy).to receive(:do_GET).twice.and_call_original
```
(no need for the `have_received` in this case) but to be verified ^^

Reply:
Yeah totally, that will work great, I usually avoid using the mock expectation that `expect().to receive` creates  because the test structure is harder to follow ... eg: with the existing code it's `setup -> execution -> expectation`, which is the same pattern for tests that aren't isolating with stubs/mocks/spies etc...  where as using the mock the steps are: `setup -> expectation -> execution` which is harder to follow as it's different to other tests:


Here's a test without any external dependencies
```ruby
# setup (this could be in a let as well)
user = User.new

# execution
result = user.name

# expectation
expect(result).to eq('Adrien')
```

Here's an example using a spy via `allow`
```ruby
# setup
user = User.new
allow(MyExternalDependency).to receive(:test)

# execution
user.name

# expectation
expect(MyExternalDependency) to have_received(:test)
```

Here's the same with the mock example via `expect ... to receive`
```ruby
# setup
user = User.new

# expectation
expect(MyExternalDependency).to receive(:test)

# execution
user.name
```

Reply:
I don't have a strong opinion on one or the other, just wanted to go through my rationale, happy to change it 👍

Reply:
Indeed, I am used to the having the expectation first with rspec mocks so that does not bother me much but I understand the rational and I agree it's easier to read/understand the way you wrote it. Let's keep it that way!

In that same spirit though, wouldn't it be a good idea to move the `allow($proxy).to receive(:do_GET).and_call_original` line inside the proxy setup `let` ? (as it's not serving any purpose for understanding of the test, it's just required to setup rspec mock so we can count the number of call at the end)

Reply:
Good idea, added in 137994a3225663d627d077eb9f50dd20fc0d5005  I moved the `sleep` too as it's part of setting up the proxy!

Reply:
Nice! and sorry but I just noticed the "let(:proxy_thread)" block is present twice and identical? can't we move it up to the top block?

Reply:
Yes, though will need to solve the problem of killing it once it's done. We have this code: `proxy_thread&.kill` to stop the proxy which will boot the proxy when calling `proxy_thread` if we moved `proxy_thread` to the top.

We could boot the proxy once for all of the tests, though we'd need to reset the stub each time.

killing the proxy could be moved to the test body but it won't run if the test runs into an exception.

Reply:
... it'll boot the proxy as the `let` is lazily loaded

Reply:
Ah yep ok, let's keep it that way then :)